packages/utils/src/index.js

```diff
@@ -1,7 +1,35 @@
 // @flow
 import type { RegisteredCache, EmotionCache, SerializedStyles } from './types'
 
-let isBrowser = typeof document !== 'undefined'
+const isBrowser = typeof document !== 'undefined'
+
+const getGlobalVersions = (context: any) => {
+  const globalVersionsKey = '__EMOTION_VERSIONS__'
+  if (!context[globalVersionsKey]) {
+    context[globalVersionsKey] = {}
+  }
+  return context[globalVersionsKey]
+}
+
+type ModuleVersions = { [key: string]: string }
+
+const versions: ModuleVersions = getGlobalVersions(isBrowser ? window : global)
+
+export function registerEmotionModule(name: string, version: string) {
+  if (versions[name] && versions[name] !== version) {
+    console.warn(
```

Comment:
It's not only about different versions (but that too) - the problem might also be caused by requiring the same version twice (from different locations at disk).

I think react-router prepared this pretty much in a way that I'd like to have this included here, take a look:
https://github.com/ReactTraining/react-router/blob/e81dfa2d01937969ee3f9b1f33c9ddd319f9e091/packages/react-router/modules/index.js#L1

For that to work exactly like in their case we'd have to have `process.env.ROLLUP_MODULE_FORMAT` available - this could potentially be added in our build tool ([preconstruct](https://github.com/preconstruct/preconstruct)). I expect supporting this would only require changes around those lines:
https://github.com/preconstruct/preconstruct/blob/120f9b6d747e985ed917a898f13d7177fd74ef73/packages/cli/src/build/rollup.ts#L164-L168

Reply:
If we want to warn based on module type as well as version, we would want to split this up:
https://github.com/preconstruct/preconstruct/blob/c1cfc0503f73a197487d6cba6ada607fe957f527/packages/cli/src/build/config.ts#L132-L153
If we allowed the messaging to be more vague, then we wouldn't need to do an explicit check of the module type. For example, we could change the code to this:
```
if (__DEV__) {
  if (versions[name]) {
    if (versions[name] === version) {
      console.warn(
        `Package ${name}@${versions[name]} is already registered, so things won't work right. This would only happen if you are loading two different builds of ${name}.`
      );
    } else {
      console.warn(
        `Package ${name}@${versions[name]} is already registered. Running ${version} may cause problems.`
      );
    }
  }
  versions[name] = version;
}
```

Reply:
The goal of this is to warn about second copy of `@emotion/core` being loaded. It doesn't really matter what was the first version being loaded and it doesn't even matter what is the second one - we only need types of both to provide a helpful error message. I would limit this to a build type error - not sure if we need to complicate this with providing also an exact version.